_drafts/code-properties-and-language-features.md

```diff
@@ -0,0 +1,50 @@
+
+# Code Properties
+
+**I think it all comes down to what properties you want your code to have, and whether a given language and paradigm make it easier or more difficult.**
+
+For example, concurrency can be a big benefit for systems that need to handle a high volume of data throughput or that are making multiple asynchronous I/O requests. You can get a little concurrency in a traditional OO language like Ruby or Java, but it's incredibly hard to use it extensively in a reliable way. As my former coworker John Gallager said about concurrency on iOS in particular, "we can safely say that all nontrivial iOS applications have at least one concurrency bug." By contrast, Elixir, Rust, and JavaScript each have a different concurrency model that makes safe concurrency much more achievable.
+
+Or consider immutability, which can simplify testing and minimize bugs from unpredictable data changes. It's possible to write immutable code in just about any language: just create new structures instead of modifying old ones. But mutable languages like Ruby and JavaScript make it incredibly difficult to *ensure* your data structures aren't modified by your code—or by third parties'. Immutability libraries like [`ice_nine`][ice_nine] and [`immutable.js`][immutable-js] can help, but they are extra hoops to jump through. By contrast, Elixir, Clojure, and Haskell are immutable by default, and it's *mutability* that requires you to jump through hoops.
+
+On the flip side, think about encapsulation, as I described above. In large systems maintained by multiple individuals or teams, encapsulation allows parts of the system to change independently of one another, minimizing "[shotgun surgery][shotgun-surgery]," where to make one change you have to modify code all over the system. I listed examples in Elixir and Redux of how you can treat data structures as encapsulated, and by convention only access or modify them via structure-specific logic. Ruby makes this easier, although the fact that the private keyword can be worked around limits how rigid it is. Other OO languages like Swift are more rigid about how they protect encapsulated data.
+
+# The Right Questions
+
+So to ask the question "should I use an OO or FP language (or style)" is to skip several steps. I think a better series of questions are: First, what properties would be beneficial for your application to have: concurrency? Immutability? Encapsulation? Once that's decided, the next question is, **in a given language, are those properties guaranteed, easy, difficult, or not realistically achievable?**
+
+Because of this, I don't think it even makes sense to say "Language X is better because it has Y properties." It's all about [your context][context]. It's more precise to say "Systems of type X benefit from property Y, and language Z makes it easier for your code to have that property."
```

Comment:
@nathanl Now that Python has async/await, perhaps it won't hurt so much. (And Twisted made that doable without the syntax sugar a decade ago.)